### ISSUE TITLE
FAILED: Docker within LXD container using 'dir' LXD storage with 'overlay' Docker storage driver

### ISSUE THREAD

**Issue by atc0005** [atc0005/ansible-role-lxd-testenv#8]
Relevant messages from `/var/log/syslog` in the LXD container:

```
Feb  7 16:49:27 ansible-ubuntu-docker systemd[1]: Starting Docker Socket for the API.
Feb  7 16:49:27 ansible-ubuntu-docker systemd[1]: Listening on Docker Socket for the API.
Feb  7 16:49:27 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/docker.service: Operation not permitted
Feb  7 16:49:27 ansible-ubuntu-docker systemd[1]: Starting Docker Application Container Engine...
Feb  7 16:49:27 ansible-ubuntu-docker dockerd[4136]: time="2019-02-07T16:49:27.181341201Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Feb  7 16:49:27 ansible-ubuntu-docker dockerd[4136]: time="2019-02-07T16:49:27.181746570Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Feb  7 16:49:27 ansible-ubuntu-docker dockerd[4136]: time="2019-02-07T16:49:27.182010542Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Feb  7 16:49:27 ansible-ubuntu-docker dockerd[4136]: time="2019-02-07T16:49:27.182234465Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Feb  7 16:49:27 ansible-ubuntu-docker dockerd[4136]: time="2019-02-07T16:49:27.183544036Z" level=error msg="'overlay' not found as a supported filesystem on this host. Please ensure kernel is new enough and has overlay support loaded." storage-driver=overlay2
Feb  7 16:49:27 ansible-ubuntu-docker dockerd[4136]: Error starting daemon: error initializing graphdriver: driver not supported
Feb  7 16:49:27 ansible-ubuntu-docker systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
Feb  7 16:49:27 ansible-ubuntu-docker systemd[1]: Failed to start Docker Application Container Engine.
Feb  7 16:49:27 ansible-ubuntu-docker systemd[1]: docker.service: Unit entered failed state.
Feb  7 16:49:27 ansible-ubuntu-docker systemd[1]: docker.service: Failed with result 'exit-code'.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: docker.service: Service hold-off time over, scheduling restart.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: Stopped Docker Application Container Engine.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: Closed Docker Socket for the API.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: Stopping Docker Socket for the API.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: Starting Docker Socket for the API.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: Listening on Docker Socket for the API.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: docker.service: Start request repeated too quickly.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: Failed to start Docker Application Container Engine.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: docker.socket: Unit entered failed state.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: docker.service: Unit entered failed state.
Feb  7 16:49:29 ansible-ubuntu-docker systemd[1]: docker.service: Failed with result 'start-limit-hit'.
```

Kernel:

`root@ansible-ubuntu-docker:~# uname -a`
```
Linux ansible-ubuntu-docker 4.4.0-142-generic #168-Ubuntu SMP Wed Jan 16 21:00:45 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

I was pretty sure that no issues were encountered when using the overlay Docker storage driver with the 4.4.0-141 kernel. Will revert to that kernel and test again.

**Comment by atc0005** [comment]
Potentially related items:

- https://github.com/docker/for-linux/issues/475
- https://github.com/containerd/containerd/issues/2772
- https://github.com/geerlingguy/ansible-role-docker/issues/97

Also related, but with a workaround:

- https://discuss.linuxcontainers.org/t/ubuntu-18-04-18-09-0-docker-on-lxd-3-0-2-3-6-is-broken/3202/6

The workaround:

`lxc config set <container> linux.kernel_modules overlay`

Output:

`ubuntu@ansiblecontrol:~$ lxc config set ansible-ubuntu-docker linux.kernel_modules overlay`

No output. Effect was immediate.

`ubuntu@ansiblecontrol:~$ ssh root@ansible-ubuntu-docker`

```
Welcome to Ubuntu 16.04.5 LTS (GNU/Linux 4.4.0-141-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage
Last login: Thu Feb  7 17:05:58 2019 from 10.245.28.1
```

`root@ansible-ubuntu-docker:~# cat /etc/docker/daemon.json `
```json
{
  "storage-driver": "overlay"
}
```

`root@ansible-ubuntu-docker:~# service docker restart`
`root@ansible-ubuntu-docker:~# systemctl status docker`

```ShellSession
● docker.service - Docker Application Container Engine
   Loaded: loaded (/lib/systemd/system/docker.service; enabled; vendor preset: enabled)
   Active: active (running) since Thu 2019-02-07 17:12:19 UTC; 7s ago
     Docs: https://docs.docker.com
 Main PID: 456 (dockerd)
    Tasks: 7
   Memory: 27.5M
      CPU: 237ms
   CGroup: /system.slice/docker.service
           └─456 /usr/bin/dockerd -H fd://

Feb 07 17:12:19 ansible-ubuntu-docker dockerd[456]: time="2019-02-07T17:12:19.807823947Z" level=info msg="ccResolverWrapper: sending new addresses to cc: [{unix:///run/containerd/containerd.sock 0  <nil>}
Feb 07 17:12:19 ansible-ubuntu-docker dockerd[456]: time="2019-02-07T17:12:19.808143450Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
Feb 07 17:12:19 ansible-ubuntu-docker dockerd[456]: time="2019-02-07T17:12:19.808393545Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc42060d620, CONNECTING" module=grpc
Feb 07 17:12:19 ansible-ubuntu-docker dockerd[456]: time="2019-02-07T17:12:19.809023111Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc42060d620, READY" module=grpc
Feb 07 17:12:19 ansible-ubuntu-docker dockerd[456]: time="2019-02-07T17:12:19.878690690Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be 
Feb 07 17:12:19 ansible-ubuntu-docker dockerd[456]: time="2019-02-07T17:12:19.915891850Z" level=info msg="Loading containers: done."
Feb 07 17:12:19 ansible-ubuntu-docker dockerd[456]: time="2019-02-07T17:12:19.969480488Z" level=info msg="Docker daemon" commit=4c52b90 graphdriver(s)=overlay version=18.09.1
Feb 07 17:12:19 ansible-ubuntu-docker dockerd[456]: time="2019-02-07T17:12:19.970081123Z" level=info msg="Daemon has completed initialization"
Feb 07 17:12:19 ansible-ubuntu-docker systemd[1]: Started Docker Application Container Engine.
Feb 07 17:12:19 ansible-ubuntu-docker dockerd[456]: time="2019-02-07T17:12:19.978350821Z" level=info msg="API listen on /var/run/docker.sock"
```

Docker starts up normally.

**Comment by atc0005** [comment]
With the changes listed previously the overlay2 Docker storage driver appears to work:

`root@ansible-ubuntu-docker:~# nano /etc/docker/daemon.json `

`root@ansible-ubuntu-docker:~# cat /etc/docker/daemon.json `
```json
{
  "storage-driver": "overlay2"
}
```

`root@ansible-ubuntu-docker:~# service docker restart`

`root@ansible-ubuntu-docker:~# systemctl status docker`
```ShellSession
● docker.service - Docker Application Container Engine
   Loaded: loaded (/lib/systemd/system/docker.service; enabled; vendor preset: enabled)
   Active: active (running) since Thu 2019-02-07 17:17:49 UTC; 3s ago
     Docs: https://docs.docker.com
 Main PID: 606 (dockerd)
    Tasks: 7
   Memory: 28.0M
      CPU: 213ms
   CGroup: /system.slice/docker.service
           └─606 /usr/bin/dockerd -H fd://

Feb 07 17:17:49 ansible-ubuntu-docker dockerd[606]: time="2019-02-07T17:17:49.150999396Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
Feb 07 17:17:49 ansible-ubuntu-docker dockerd[606]: time="2019-02-07T17:17:49.151294294Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc4206035f0, CONNECTING" module=grpc
Feb 07 17:17:49 ansible-ubuntu-docker dockerd[606]: time="2019-02-07T17:17:49.151815864Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc4206035f0, READY" module=grpc
Feb 07 17:17:49 ansible-ubuntu-docker dockerd[606]: time="2019-02-07T17:17:49.230731552Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be 
Feb 07 17:17:49 ansible-ubuntu-docker dockerd[606]: time="2019-02-07T17:17:49.259111611Z" level=info msg="Loading containers: done."
Feb 07 17:17:49 ansible-ubuntu-docker dockerd[606]: time="2019-02-07T17:17:49.280515931Z" level=warning msg="Not using native diff for overlay2, this may cause degraded performance for building images: fa
Feb 07 17:17:49 ansible-ubuntu-docker dockerd[606]: time="2019-02-07T17:17:49.281117720Z" level=info msg="Docker daemon" commit=4c52b90 graphdriver(s)=overlay2 version=18.09.1
Feb 07 17:17:49 ansible-ubuntu-docker dockerd[606]: time="2019-02-07T17:17:49.281414797Z" level=info msg="Daemon has completed initialization"
Feb 07 17:17:49 ansible-ubuntu-docker systemd[1]: Started Docker Application Container Engine.
Feb 07 17:17:49 ansible-ubuntu-docker dockerd[606]: time="2019-02-07T17:17:49.290515858Z" level=info msg="API listen on /var/run/docker.sock"
```

**Comment by atc0005** [comment]
Included in #14.
